src/crimson/osd/watch.cc

```diff
@@ -169,10 +182,38 @@ seastar::future<> Notify::maybe_send_completion()
       CEPH_WATCH_EVENT_NOTIFY_COMPLETE,
       empty,
       client_gid);
-    reply->set_data(bl);
+    ceph::bufferlist reply_bl;
+    {
+      std::list<std::pair<uint64_t,uint64_t>> missed;
+      boost::insert(
+	missed, std::begin(missed),
+	timedout_watchers | boost::adaptors::transformed([] (auto w) {
+          return std::make_pair(w->get_watcher_gid(), w->get_cookie());
+        }));
+      ceph::encode(notify_replies, reply_bl);
+      ceph::encode(missed, reply_bl);
+    }
+    reply->set_data(std::move(reply_bl));
+    if (!timedout_watchers.empty()) {
+      reply->return_code = -ETIMEDOUT;
+    }
     return conn->send(std::move(reply));
   }
   return seastar::now();
 }
 
+void Notify::do_timeout()
+{
+  logger().debug("{} complete={}", __func__, complete);
+  if (complete) {
+    return;
+  }
+  decltype(watchers) timedout_watchers;
+  std::swap(watchers, timedout_watchers);
+  for (auto& watcher : timedout_watchers) {
+    watcher->cancel_notify(ninfo.notify_id);
+  }
+  std::ignore = maybe_send_completion(std::move(timedout_watchers));
```

Comment:
also, i am confused, why bother moving `watchers` and passing a move copy of it to `maybe_send_completion()`? we could just let `maybe_send_completion()` transform and encode `watchers` if we pass a `true` `timedout` parameter to it, no?

Reply:
Yes, but if so we should rename as it would start doing more than just sending. Actually, even now canceling the timer fits into _doing more than_; though,altering the `watchers` would be even more.

Reply:
makes sense to me.